File: atcoder/abc276_e.cpp
// https://atcoder.jp/contests/abc276/tasks/abc276_e
#include<iostream>
// #include<algorithm>
// #include<string>
// #include<numeric>
// #include<vector>
// #include<map>
// #include<tuple>
// #include<set>
// #include<queue>
// #include<deque>
// #include<regex>
// #include<bitset>
// #include<iomanip>
// #include<complex>
// #include<stack>
// #include<functional>

#include<bits/stdc++.h>
using namespace std;
#include <atcoder/all>
using namespace atcoder;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define drep(i,n) for(int i = (n-1); i >= 0; i--)
#define all(v) (v).begin(),(v).end()
#define maxs(x,y) (x = max(x,y))
#define mins(x,y) (x = min(x,y))
template <class T> bool chmax(T &a, const T &b) { if (a < b) { a = b; return 1; } return 0; }
template <class T> bool chmin(T &a, const T &b) { if (b < a) { a = b; return 1; } return 0; }
template <class T> T gcd(T a, T b) { return b ? gcd(b, a % b) : a; }
template <class T> T lcm(T a, T b) { return a/gcd(a,b)*b; }
typedef pair<int, int> P;
typedef long long ll;
const int INF = 1001001001;
const ll LINF = 1001002003004005006ll;
const ll MOD = 1e9+7;

const int di[] = {-1, 0, 1, 0};
const int dj[] = { 0,-1, 0, 1};

int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
    int h, w;
	cin >> h >> w;
    vector<string> c(h);
    rep(i,h) cin >> c[i];
    int n = h*w;
    dsu uf(n);
    rep(v,n) {
        int i = v/w, j = v%w;
        if (c[i][j]!='.') continue;
        if (i+1<h && c[i+1][j]=='.') uf.merge(v, v+w);
        if (j+1<w && c[i][j+1]=='.') uf.merge(v, v+1);
    }
    set<int> st;
    int cnt = 0;
    rep(i,h) rep(j,w) if(c[i][j]=='S') {
        rep(v, 4) {
            int ni = i + di[v], nj = j + dj[v];
            if (ni<0 || ni>=h) continue;
            if (nj<0 || nj>=w) continue;
            if (c[ni][nj]!='.') continue;
            st.insert(uf.leader(ni*w+nj));
            cnt++;
        }
    }

    if (st.size()!=cnt) {
    	cout << "Yes" << "\n";
    } else {
        cout << "No" << "\n";
    }
	return 0;
}



// const int MAX_H = 1000005;
// vector<string> c(MAX_H);
// vector<bool> gone(MAX_H);
// bool ans = false;
// int h, w;
// int si, sj;
// vector<int> vi{ 0, 0, -1, +1 };
// vector<int> vj{ -1, +1, 0, 0 };

// void dfs(int i, int j, int n) {
//     // cout << i << ' ' << j << endl;
//     if (i==si && j==sj && n>=4) {
//         ans = true;
//         return;
//     }
//     // cout << i << ' ' << j << endl;
//     rep(idx, 4) {
//         int ni = i + vi[idx];
//         // cout << i << ' ' << ni << ' ' << vi[idx] << endl;
//         if (ni<0 || ni>=h) continue;
//         int nj = j + vj[idx];
//         // cout << j << ' ' << nj << ' ' << vj[idx] << endl;
//         if (nj<0 || nj>=w) continue;
//         // cout << 'c' << endl;
//         // cout << c[ni][nj] << endl;
//         if (c[ni][nj]=='#') continue;
//         if (gone[ni*w+nj]) continue;
//         gone[ni*w+nj] = true;
//         // cout << gone[ni*w+nj] << endl;
//         dfs(ni, nj, n+1);
//         gone[ni*w+nj] = false;
//     }
//     return;
// }

// int main() {
//     cin.tie(nullptr);
//     ios::sync_with_stdio(false);
// 	cin >> h >> w;
//     // cout << h << ' ' << w << endl;
//     // vector<string> c(h);
//     rep(i,h) cin >> c[i];
//     // cout << c[0] << endl;
//     vector<bool> gone(h*w, false);
//     // cout << gone[0][0] << endl;
//     rep(i,h) rep(j,w) if(c[i][j]=='S') {
//         si = i, sj = j;
//         break;
//     }
//     // cout << si << ' ' << sj << endl;
//     dfs(si, sj, 0);
//     if (ans) {
//     	cout << "Yes" << "\n";
//     } else {
//         cout << "No" << "\n";
//     }
// 	return 0;
// }

File: atcoder/abc276_e.py
# https://atcoder.jp/contests/abc276/tasks/abc276_e


import sys
# input = sys.stdin.buffer.readline
# sys.setrecursionlimit(10 ** 7)
from itertools import combinations

class UnionFind():
    """ Union-Find木の実装（ランクあり） """
    def __init__(self, n):
        self.n = n
        self.parent = [-1]*(n+1)    # parent 親 マイナスだったら根で、絶対値が集合の数
        self.rank   = [0]*(n+1)     # rank 深さ

    def root(self, x):
        """ 木の根を求める """
        if self.parent[x] < 0:      # xが根
            return x
        else:
            self.parent[x] = self.root(self.parent[x])  # 経路圧縮
            return self.parent[x]
            
    def unite(self, x, y):
        """ xとyの属する集合を併合 """
        x = self.root(x)
        y = self.root(y)
        if x == y:
            return 0
        elif self.rank[x] > self.rank[y]:
            self.parent[x] += self.parent[y]
            self.parent[y] = x
        else:
            self.parent[y] += self.parent[x]
            self.parent[x] = y
            if self.rank[x] == self.rank[y]:
                self.rank[y] += 1

    def isSame(self, x, y):
        """ xとyが同じ集合に属するか否か """
        return self.root(x) == self.root(y) or x==y
    
    def size(self, x):
        """ xと同じ集合に属する数 """
        return -self.parent[self.root(x)]

vx = [-1, 1, 0, 0]
vy = [ 0, 0, 1,-1]
H, W = map(int, input().split())
C = [input() for _ in range(H)]
# print(C)

for i in range(H):
    for j in range(W):
        if C[i][j]=='S':
            sx = i
            sy = j
            break
# print(si, sj)
cand = []
for j in range(4):
    nx = sx + vx[j]
    if nx<0 or nx>=H: continue
    ny = sy + vy[j]
    if ny<0 or ny>=W: continue
    if C[nx][ny]!='.': continue
    cand.append(nx*W+ny)
# print(cand)

uf = UnionFind(H*W)

for i in range(H*W):
    x, y = divmod(i, W)
    if C[x][y]!='.': continue
    for j in range(4):
        nx = x + vx[j]
        if nx<0 or nx>=H: continue
        ny = y + vy[j]
        if ny<0 or ny>=W: continue
        if C[nx][ny]!='.': continue
        uf.unite(i, nx*W+ny)

ans = "No"
for i, j in combinations(cand, 2):
    # print(i, j)
    if uf.isSame(i, j):
        ans = "Yes"
        break
print(ans)




# TLE
# import sys
# # input = sys.stdin.buffer.readline
# sys.setrecursionlimit(10 ** 7)

# H, W = map(int, input().split())
# C = [input() for _ in range(H)]
# # print(C)

# for i in range(H):
#     for j in range(W):
#         if C[i][j]=='S':
#             si = i
#             sj = j
#             break
# # print(si, sj)

# vi = [-1, 1, 0, 0]
# vj = [ 0, 0, 1,-1]

# ans = "No"

# gone = [[False]*W for _ in range(H)]

# def dfs(i, j, n):
#     global ans
#     if i==si and j==sj and n>=4:
#         ans = "Yes"
#         return
#     for idx in range(4):
#         ni = i + vi[idx]
#         if ni<0 or ni>=H: continue
#         nj = j + vj[idx]
#         if nj<0 or nj>=W: continue
#         if C[ni][nj]=='#': continue
#         if gone[ni][nj]: continue
#         gone[ni][nj] = True
#         dfs(ni, nj, n+1)
#         gone[ni][nj] = False
#     return

# dfs(si, sj, 0)
# print(ans)

